warc_metadata_sidecar.py

```diff
@@ -0,0 +1,261 @@
+                   }
+    return result_dict
+
+
+def find_language(bytes_payload):
+    """Find the language of the payload using pycld2."""
+    is_reliable, text_bytes_found, details = cld2.detect(bytes_payload.decode('utf-8', 'replace'),
+                                                         bestEffort=True)
+    new_list = []
+    # 'details' seems to always return 3, if the language is 'Unknown' we don't need to list it.
+    for item in details:
+        if 'Unknown' not in item:
+            new_list.append({
+                'name': item[0],
+                'code': item[1],
+                'text-covered': item[2],
+                'score': item[3]})
+    if len(new_list):
+        lang_cld = {'reliable': is_reliable,
+                    'text-bytes': text_bytes_found,
+                    'languages': new_list}
+        return lang_cld
+    else:
+        return None
+
+
+def create_warcinfo_payload(payload, operator, publisher, new_file, version):
+    payload.seek(0)
+    info_payload = payload.read()
+    warcinfo = info_payload.decode("utf-8")
+    warc_list = re.split('\r\n|: ', warcinfo)
+    hostname = socket.gethostname()
+    warc_doc = 'http://bibnum.bnf.fr/WARC/WARC_ISO_28500_version1_latestdraft.pdf'
+    warcinfo_payload = {'software': 'warc-metadata-sidecar/' + version,
+                        'hostname': hostname,
+                        'ip': socket.gethostbyname(hostname),
+                        'conformsTo': warc_doc,
+                        'description': 'WARC metdata sidecar for ' + new_file,
+                        'publisher': publisher}
+    if operator:
+        warcinfo_payload['operator'] = operator
+    if 'isPartOf' in warc_list:
+        list_index = warc_list.index('isPartOf')
+        is_part_of = warc_list[list_index + 1]
+        warcinfo_payload['isPartOf'] = is_part_of
+
+    return warcinfo_payload
+
+
+def metadata_sidecar(archive_dir, warc_file, operator=None,
+                     publisher='University of North Texas - Digital Projects Unit'):
+    start = time.time()
+    version = pkg_resources.require("warc-metadata-sidecar")[0].version
+
+    if not os.path.isdir(archive_dir):
+        os.mkdir(archive_dir)
+
+    logging.basicConfig(
+        filename=os.path.join(archive_dir, 'sidecar.log'),
+        level=logging.INFO,
+        format='%(asctime)s %(levelname)s %(message)s',
+    )
+    logging.getLogger(__name__)
+    logging.info('Logging WARC metadata record information for %s', warc_file)
+
+    # Create sidecar filename, adding 'meta' as extension.
+    new_file = os.path.basename(warc_file)
+    meta_file = re.sub('warc(\.gz)?$', 'warc.meta.gz', new_file)
+    logging.info('Creating sidecar %s', meta_file)
+    warc_file_path = os.path.join(archive_dir, meta_file)
+
+    fido = ExtendFido()
+    mime_title = 'Identified-Payload-Type:'
+    puid_title = 'Preservation-Identifier:'
+    charset_title = 'Charset-Detected:'
+    language_title = 'Languages-cld2:'
```

Comment:
Could move setting the title strings outside of the function (to the top of the file) and capitalize the names for constants. This would allow them to be used in the tests if desired:

```   
MIME_TITLE = 'Identified-Payload-Type:'
PUID_TITLE = 'Preservation-Identifier:'
CHARSET_TITLE = 'Charset-Detected:'
LANGUAGE_TITLE = 'Languages-cld2:'
```